compiler/aarch64/codegen/OMRTreeEvaluator.cpp

```diff
@@ -6400,10 +6400,74 @@ OMR::ARM64::TreeEvaluator::arraytranslateAndTestEvaluator(TR::Node *node, TR::Co
 
 TR::Register *
 OMR::ARM64::TreeEvaluator::arraytranslateEvaluator(TR::Node *node, TR::CodeGenerator *cg)
-	{
-	// TODO:ARM64: Enable TR::TreeEvaluator::arraytranslateEvaluator in compiler/aarch64/codegen/TreeEvaluatorTable.hpp when Implemented.
-	return OMR::ARM64::TreeEvaluator::unImpOpEvaluator(node, cg);
-	}
+   {
+   // tree looks as follows:
+   // arraytranslate
+   //    (0) input ptr
+   //    (1) output ptr
+   //    (2) translation table (dummy)
+   //    (3) stop character (terminal character, either 0xff00ff00 (ISO8859) or 0xff80ff80 (ASCII)
+   //    (4) input length (in elements)
+   //    (5) stopping char (dummy)
+   //
+   // Number of translated elements is returned
+
+   TR::Compilation *comp = cg->comp();
+
+   TR_ASSERT_FATAL(!node->isSourceByteArrayTranslate(), "Source is byte[] for arraytranslate");
+   TR_ASSERT_FATAL(node->isTargetByteArrayTranslate(), "Target is char[] for arraytranslate");
+   TR_ASSERT_FATAL(node->getChild(3)->getOpCodeValue() == TR::iconst && node->getChild(3)->getInt() == 0x0ff00ff00, "Non-ISO8859 stop character for arraytranslate");
+
+   static bool verboseArrayTranslate = (feGetEnv("TR_verboseArrayTranslate") != NULL);
+   if (verboseArrayTranslate)
+      {
+      fprintf(stderr, "arrayTranslateTRTO255: %s @ %s\n",
+         comp->signature(),
+         comp->getHotnessName(comp->getMethodHotness())
+         );
+      }
+
+   TR::Register *inputReg = cg->gprClobberEvaluate(node->getChild(0));
+   TR::Register *outputReg = cg->gprClobberEvaluate(node->getChild(1));
+   TR::Register *inputLenReg = cg->gprClobberEvaluate(node->getChild(4));
+   TR::Register *outputLenReg = cg->allocateRegister();
```

Comment:
Children 2,3,5 still need to be evaluated even if they are "dummy".

Reply:
Arraytranslate evaluators of p and x platforms leave those dummy children unevaluated.

- p: https://github.com/eclipse-omr/omr/blob/be39f059c4e1d1ee4d3448511780ebe66c7551b8/compiler/p/codegen/OMRTreeEvaluator.cpp#L5087-L5090
- x: https://github.com/eclipse-omr/omr/blob/be39f059c4e1d1ee4d3448511780ebe66c7551b8/compiler/x/codegen/OMRTreeEvaluator.cpp#L3462-L3465

Reference counts of all the child nodes are decremented.

Reply:
Generally, I think you can only get away with that if they are constants.  Are they?

Reply:
Yes, those dummy nodes are constants.